### PR TITLE
Feature/implementClientModule

### PR DESCRIPTION
**Pull request by pavloshumylo**

### CODE REVIEW

File: client/src/main/java/com/foursquare/Application.java [hunk 1]
```diff
@@ -0,0 +1,12 @@
+package com.foursquare;
+
+import org.springframework.boot.SpringApplication;
+import org.springframework.boot.autoconfigure.SpringBootApplication;
+
+@SpringBootApplication
+public class Application {
+
+    public static void main(String[] args) {
```
Comment (serg1236) [inline client/src/main/java/com/foursquare/Application.java]: Client is a library. Its ok to create app for manual testing but not ok for production.

File: client/src/main/java/com/foursquare/endpoints/FoursquareClientEndpoints.java [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+
+public interface FoursquareClientEndpoints {
+
+    CompletableFuture<ResponseEntity<SearchResponseDto>> invokeSearchEndPoint(String city, String query, String limit, String accessToken);
+    CompletableFuture<ResponseEntity<ResponseEntity>> invokeRegistrationEndPoint(User user);
+    CompletableFuture<ResponseEntity<ResponseEntity>> invokeSaveEndpoint(String fsId, String accessToken);
+    CompletableFuture<ResponseEntity<ResponseEntity>> invokeDeleteEndpoint(String fsId, String accessToken);
+    CompletableFuture<ResponseEntity<List<Venue>>> invokeGetEndpoint(String accessToken);
```
Comment (serg1236) [inline client/src/main/java/com/foursquare/endpoints/FoursquareClientEndpoints.java]: No response entities. Body should be unwrapped inside of the client.

Comment (serg1236) [inline client/src/main/java/com/foursquare/endpoints/FoursquareClientEndpoints.java]: Think about other ways to pass the access token. (ITS NOT AN ISSUE, JUST THINK:)

File: client/src/main/java/com/foursquare/endpoints/impl/FoursquareClientEndpointsImpl.java [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+import org.springframework.util.concurrent.ListenableFutureCallback;
+import org.springframework.web.client.AsyncRestTemplate;
+
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+
+@Component
+public class FoursquareClientEndpointsImpl implements FoursquareClientEndpoints {
+
+    @Value("${foursquare.endPointHost:http://localhost:8080/}")
```
Comment (serg1236) [inline client/src/main/java/com/foursquare/endpoints/impl/FoursquareClientEndpointsImpl.java]: If you use protocol here, it is not a host already, it is base url or something like that.

Comment (serg1236) [reply client/src/main/java/com/foursquare/endpoints/impl/FoursquareClientEndpointsImpl.java]: And I would not put the trailing slash here:)

File: client/src/main/java/com/foursquare/endpoints/impl/FoursquareClientEndpointsImpl.java [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+
+        return buildCompletableFuture(listenableFuture);
+    }
+
+    private static <T> CompletableFuture<T> buildCompletableFuture(ListenableFuture<T> listenableFuture) {
+        CompletableFuture<T> completable = new CompletableFuture<T>();
+
+        listenableFuture.addCallback(new ListenableFutureCallback<T>() {
+            @Override
+            public void onSuccess(T result) { completable.complete(result); }
```
Comment (serg1236) [inline client/src/main/java/com/foursquare/endpoints/impl/FoursquareClientEndpointsImpl.java]: It will be `result.getBody()` after you fix the interface.

File: client/src/main/java/com/foursquare/endpoints/impl/FoursquareClientEndpointsImpl.java [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+                new ParameterizedTypeReference<List<Venue>>() {
+                },
+                host,
+                accessToken);
+
+        return buildCompletableFuture(listenableFuture);
+    }
+
+    private static <T> CompletableFuture<T> buildCompletableFuture(ListenableFuture<T> listenableFuture) {
+        CompletableFuture<T> completable = new CompletableFuture<T>();
```
Comment (serg1236) [inline client/src/main/java/com/foursquare/endpoints/impl/FoursquareClientEndpointsImpl.java]: Implement `onCancel` for the CompletableFuture. It should cancel Listenable one.

File: client/src/main/java/com/foursquare/endpoints/impl/FoursquareClientEndpointsImpl.java [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+            @Override
+            public void onFailure(Throwable t) {
+                completable.completeExceptionally(t);
+            }
+        });
+
+        return completable;
+    }
+
+    public void setHost(String host) {
```
Comment (serg1236) [inline client/src/main/java/com/foursquare/endpoints/impl/FoursquareClientEndpointsImpl.java]: We shouldn't give ability to change it.

File: client/src/main/resources/application.yml [hunk 7]
```diff
@@ -0,0 +1,4 @@
+foursquare:
+  endPointHost: http://localhost:8080/
+server:
+  port: ${SERVER_PORT:8080}
```
Comment (serg1236) [inline client/src/main/resources/application.yml]: why do you need this? Isn't 8080 default?

File: client/src/main/java/com/foursquare/endpoints/FoursquareClientEndpoints.java [hunk 8]
```diff
@@ -0,0 +1,18 @@
...
+
+import com.foursquare.dto.SearchResponseDto;
+import com.foursquare.entity.User;
+import com.foursquare.entity.Venue;
+import org.springframework.http.ResponseEntity;
+
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+
+public interface FoursquareClientEndpoints {
```
Comment (serg1236) [inline client/src/main/java/com/foursquare/endpoints/FoursquareClientEndpoints.java]: `ClientEndpoints` means "endpoints of the client" which is at least strange:) It's just a `FoursquareClient`.

File: client/src/test/java/com/foursquare/endpoints/FoursquareClientEndpointsImplTest.java [hunk 9]
```diff
@@ -0,0 +1,124 @@
...
+
+    @Autowired
+    private FoursquareClientEndpointsImpl foursquareClientEndpoints;
+
+    @Rule
+    public WireMockRule wireMockRule = new WireMockRule(wireMockConfig().dynamicPort());
+
+    @Before
+    public void init() {
+        foursquareClientEndpoints.setHost("http://localhost:"+wireMockRule.port()+"/");
```
Comment (serg1236) [inline client/src/test/java/com/foursquare/endpoints/FoursquareClientEndpointsImplTest.java]: Aha...that's why you need `setHost`. We should not add extra methods just to test easier. But I agree it is not simple to pass the dynamic port. There is a way to pass it to properties but it is not simple... For now just use `org.springframework.util.ReflectionUtils`

File: client/src/test/java/com/foursquare/endpoints/FoursquareClientEndpointsImplTest.java [hunk 10]
```diff
@@ -0,0 +1,124 @@
...
+    }
+
+    @Test
+    public void testInvokeSearchEndPoint_ShouldReturnCompletableFutureWithSearchResponseDto() throws IOException, ExecutionException, InterruptedException {
+        InputStream inputStreamWithSearchResponseDtoFirst = getClass().getClassLoader()
+                .getResourceAsStream("testData/search_response_dto.json");
+
+        InputStream inputStreamWithSearchResponseDtoSecond = getClass().getClassLoader()
```
Comment (serg1236) [inline client/src/test/java/com/foursquare/endpoints/FoursquareClientEndpointsImplTest.java]: No need to load it twice. Use one only.

File: client/src/test/java/com/foursquare/endpoints/FoursquareClientEndpointsImplTest.java [hunk 11]
```diff
@@ -0,0 +1,124 @@
+package com.foursquare.endpoints;
```
Comment (serg1236) [inline client/src/test/java/com/foursquare/endpoints/FoursquareClientEndpointsImplTest.java]: We'll also need tests for `onCancel` and negative tests.

**Review by serg1236 — CHANGES_REQUESTED** [review]

File: client/src/main/java/com/foursquare/endpoints/FoursquareClient.java [hunk 12]
```diff
@@ -0,0 +1,17 @@
...
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+
+public interface FoursquareClient {
+
+    CompletableFuture<SearchResponseDto> invokeSearchEndPoint(String city, String query, String limit, String accessToken);
+    CompletableFuture<Void> invokeRegistrationEndPoint(User user);
+    CompletableFuture<Void> invokeSaveEndpoint(String fsId, String accessToken);
+    CompletableFuture<Void> invokeDeleteEndpoint(String fsId, String accessToken);
+    CompletableFuture<List<Venue>> invokeGetEndpoint(String accessToken);
```
Comment (serg1236) [inline client/src/main/java/com/foursquare/endpoints/FoursquareClient.java]: This naming (my bad, I missed it) makes me cry. Imagine that you are a user of this service. You inject the client to your service and go inside to see what you can do. And what do you see? invokeBlahEndpoint, invokeBlahBlahEndpoint, ... And you like: "WTF? Why do I need to invoke something? What will it cause? What I delete if I invoke delete endpoint? Or I do not delete anything?".

Comment (serg1236) [reply client/src/main/java/com/foursquare/endpoints/FoursquareClient.java]: Try to use simple names. Same as you use on a service layers (maybe with some additional details). For instance: "registerNewUser", "registerUser", "removeVenue", "getAllVenues", "getAllVenuesForUser" etc.

File: client/src/test/java/com/foursquare/endpoints/FoursquareClientImplTest.java [hunk 13]
```diff
@@ -0,0 +1,245 @@
...
+
+            Assert.fail();
+        }
+        catch (ExecutionException ex) {
+            assertTrue(ex.getCause() instanceof HttpClientErrorException);
+        }
+    }
+
+    @Test(expected = CancellationException.class)
+    public void testbuildCompletableFuture_Cancel_ShouldThrowCancellationException() throws InterruptedException, ExecutionException {
```
Comment (serg1236) [inline client/src/test/java/com/foursquare/endpoints/FoursquareClientImplTest.java]: Are we able to check somehow if ListenableFuture was canceled also?

**Review by serg1236 — CHANGES_REQUESTED** [review]
Two comments. Try to create a better test and fix the naming.

**Review by sdakhnii-eagle — APPROVED** [review]
